### ISSUE TITLE
alternating optimization and joint optimization are not equal

### ISSUE THREAD

**Issue by gaopeng-eugene** [chrischoy/fully-differentiable-deep-ndf-tf#2]

**Comment by gaopeng-eugene** [comment]
In the ICCV paper, the author only proposed alternating optimisation trained with resilient SGD. Joint optimization is definitely not the right solution. If they author proposed joint optimization in the ICCV paper, the contribution is minor and will not be ICCV best paper.

**Comment by chrischoy** [comment]
Thank you for your word of wisdom. Duly noted.

**Comment by chrischoy** [comment]
Impeccable reasoning!
Joint optimization is wrong because, otherwise, they wouldn't have won the award!

**Comment by gaopeng-eugene** [comment]
In the iccv paper supplement, they prove alternating optimization for pi is convex and guarantee converge to a unique optimal by several iterations. The info Gan is very similiar with random forest and decision tree. However, when you use joint optimization nothing exist. Unless you can give a prove.  Please go to author’s homepage and read the provement.

**Comment by gaopeng-eugene** [comment]
Iteration for pi is convex not the whole system.

**Comment by gaopeng-eugene** [comment]
Another reason I found this paper is interesting is dynamic routing in capsule net and follow-up paper EM optimization. It seems that this two papers have some link.

**Comment by gaopeng-eugene** [comment]
Another to prove joint optimization is wrong is when you visulize the distribution in the leaf. There is no patterns because it’s only a feature representation which do not contain any semantic meaning. However, when you train by EM. Same class will follow similiar routing path. The leaf distribution is more interpretable just because past experience show decision tree leaf is very easy to understand.

**Comment by gaopeng-eugene** [comment]
Anyway, thank you so much for your work. It’s very beneficial for me to understand the whole idea.

**Comment by gaopeng-eugene** [comment]
http://www.dsi.unive.it/~srotabul/files/publications/CVPR2014a_supp.pdf

**Comment by gaopeng-eugene** [comment]
Some typo 
Info Gan => information gain

**Comment by chrischoy** [comment]
Seems like you just read the analysis but yet understood what are the disadvantages of EM and why it requires convexity analysis.

First, EM algorithm, or alternating optimization in general, suffers from slow convergence. This is because the other variables that are not optimized slow down the convergence of the other variable.

Second, alternating optimization is not guaranteed to converge to the optimum. This is why more analysis is required for alternating optimization.

I explained the same thing in the readme. So please read before you make random comments that do not make much sense.

**Comment by gaopeng-eugene** [comment]
Thank you so much for your answer. Have you ever compared joint optimisation with alternating optimisation ? If joint optimisation is the right answer, why the author chose the much more complex and difficult to implement alternating optimisation?

**Comment by gaopeng-eugene** [comment]
Dear Gao,
  thank you for your interest in the work. You can in principle also optimize jointly the leaf statistics. 
As far as I remember, we tried that too, but the solution reported in the paper was more effective.
But could also be a matter of hyperparameter tuning (e.g., different learning rates for leaf statistics, etc).

Best,

Samuel

**Comment by gaopeng-eugene** [comment]
The author respond me just now saying that joint optimization is OK. But they only report EM method because its more effective.

**Comment by chrischoy** [comment]
Okay, I'm not working on this project. Why don't you implement their method and compare? You just have to remove the softmax variables from the optimization and optimize them separately using the update rule they proposed. It wouldn't be that difficult and this way, you can actually contribute to the society. (:

**Comment by gaopeng-eugene** [comment]
Sure, I will test it on ROI region with joint optimisation. Again, big thanks for the implementation. Looking forward to see your interesting work. 2D-R2N2 is very good 😊

**Comment by gaopeng-eugene** [comment]
Do you know any paper proposing boost differentiable decision forests? Basically, the newly added tree will focus on misclassified sample ?
